### PR TITLE
Add an agent injecting mutating webhook to the traffic manager.

### PR DESCRIPTION
**Pull request by thallgren**
## Description

This PR adds a mutating webhook that injects a traffic-agent to
pods that have the following annotation:

    telepresence.getambassador.io/traffic-agent: enabled

The installer will henceforth ignore deployments that have this
annotation and consider them to already have an agent installed.

The webhook is using `kates.List()` to list Service objects which
means that the cluster must be configured to allow the traffic-manager
to do that. A sample `k8s/agent-injector-rbac.yaml` is included that
provides the needed credentials.

A lot of code used by the agent installer is reused in the injector
and was therefore moved into the `agentinstall` package that they
both now share.
## Checklist

 - [ ] I made sure to update `./CHANGELOG.md`.
 - [ ] I made sure to either submit a docs PR, or tell Matt about the necessary documentation changes.
 - [ ] My change is adequately tested.
 - [ ] I updated `DEVELOPING.md` with any any special dev tricks I had to use to work on this code efficiently.

### CODE REVIEW

File: k8s/agent-injector-rbac.yaml
Comment (donnyyung) [inline k8s/agent-injector-rbac.yaml]: I think this rbac is actually something that is needed for the operator who will be deploying the traffic-manager and not users (who won't need this role attached to their account and shouldn't need these permissions).

Comment (thallgren) [reply k8s/agent-injector-rbac.yaml]: Doh, that comment is off. I'll change it.

Comment (thallgren) [reply k8s/agent-injector-rbac.yaml]: Hmm, it's already changed. You need to refresh your browser :-)

**Comment by LukeShu** [comment]
If `envconfig.go` is moving from `manager/` to `manager/managerutil/`, then `envconfig_test.go` should probably move too?

**Comment by donnyyung** [comment]
I'm reliably getting: ```telepresence: error: Intercept "3b3c482d-4de9-4145-a343-f7d8fff0e3aa:echo-easy": Failed to establish intercept: intercept in error state AGENT_ERROR: no valid license & cannot connect to Ambassador Cloud for license check``` when using this branch with the Smart Agent.

Looking at the config, 
```
time="2021-06-10 15:25:43" level=info msg="config = {Config:{Name:echo-easy Namespace:default PodIP:10.42.0.45 AgentPort:9900 AppMounts:/tel_app_mounts AppPort:8080 ManagerHost:traffic-manager.ambassador ManagerPort:8081} EnvoyLogLevel:info EnvoyBaseID:1 UseHTTP2:true TerminatingTLSSecret: OriginatingTLSSecret: License: LicenseHost: LicensePublicKeys:{KeyMap:map[auth.datawire.io:-----BEGIN PUBLIC KEY----------END PUBLIC KEY----- beta-auth.datawire.io:-----BEGIN PUBLIC KEY----------END PUBLIC KEY-----]} ClusterID: CanConnectCloud:false SystemAHost: SystemAPort:}" func=main.Main file="github.com/datawire/telepresence2/proprietary/cmd/app-sidecar/sidecar.go:178" PID=1
```

Looking at the config, SystemAHost and SystemAPort (which the agent talks to the manager to get), aren't being correctly populated. I'll keep looking but wanted to raise the issue here in case you might know what in the changes might have affected this.

-- Edit --
I added a comment with what I believe to be the cause of the issue.

File: cmd/traffic/cmd/manager/service.go
Comment (donnyyung) [inline cmd/traffic/cmd/manager/service.go:54]: I think this is the issue with the smart agent bc this function uses the config to get the System A info: https://github.com/telepresenceio/telepresence/blob/release/v2/cmd/traffic/cmd/manager/service.go#L100-L102

**Comment by LukeShu** [comment]
Why not straight-up use the existing install "actions" system?

**Comment by LukeShu** [comment]
"Add an agent injecting mutating webhook to the traffic manager" is difficult to review... it should probably be multiple commits.

File: cmd/traffic/cmd/manager/internal/mutator/agent_injector.go
Comment (LukeShu) [inline cmd/traffic/cmd/manager/internal/mutator/agent_injector.go]: IMO, importing things as `v1` is an anti-pattern.  Say what it's v1 of! i.e. `corev1`

File: pkg/install/container.go
Comment (LukeShu) [inline pkg/install/container.go]: That's a lot of args.  Maybe it should take a single struct?

At least wrap this to be 1 arg per line.

File: cmd/traffic/cmd/manager/managerutil/envconfig.go
Comment (LukeShu) [inline cmd/traffic/cmd/manager/managerutil/envconfig.go]: I don't like having short envconfig field names like this.  If we thought it was worth it to put `MANAGER_` in the env-var name, then surely it's worth it to put `Manager` in the field name.

File: cmd/traffic/cmd/manager/internal/mutator/agent_injector.go
Comment (LukeShu) [inline cmd/traffic/cmd/manager/internal/mutator/agent_injector.go]: Consider doing this in `LoadEnv`

File: build/admission_controller_tls.go
Comment (LukeShu) [inline build/admission_controller_tls.go:1]: So... when does this need to get run?  Because nothing's running it right now.

File: cmd/traffic/cmd/manager/systema.go
Comment (LukeShu) [inline cmd/traffic/cmd/manager/systema.go]: This should go further down, with the rest of the other telepresenceio imports.

File: cmd/traffic/cmd/manager/service.go
Comment (LukeShu) [inline cmd/traffic/cmd/manager/service.go]: This should go further down, with the rest of the telepresence.io imports.

**Comment by LukeShu** [comment]
I want to say: I'm making a lot of style-related comments.  The point of them isn't to nitpick the style; I'm commenting when it's something that makes review harder and is costing me time.  For instance, I wanted to know what changed between `install_actions.go` and `install/container.go`, but that turned out to be needlessly difficult because `corev1` was renamed to `v1` and a bunch of brackets changed.  These things are style nitpicks that  on their own wouldn't be worth mentioning.  But they make the diff be 90% noise, so it's difficult to see what's actually changing.

File: pkg/client/config.go
Comment (LukeShu) [inline pkg/client/config.go:341]: This should be called out as a "change" in the changelog (with a note about why it changed).

File: pkg/client/envconfig.go
Comment (LukeShu) [inline pkg/client/envconfig.go]: Supporting this is probably worth mentioning in the changelog.

**Review by LukeShu — CHANGES_REQUESTED** [review]
Please make this easier to review.  I spent all morning on it, and didn't get through it.

File: pkg/client/envconfig.go
Comment (thallgren) [reply pkg/client/envconfig.go]: Perhaps, but it isn't a change really. We've always supported that env using an `os.GetEnv()` in the installer.

File: build/admission_controller_tls.go
Comment (thallgren) [reply build/admission_controller_tls.go:1]: This utility was created before I automated all of it in the installer. I thought it worth keeping around because it might help if we need to generate certs from the helm-chart.

**Comment by thallgren** [comment]
@LukeShu there are two reasons why I didn't use the installer "actions":

1. There's no need to `do`/`undo`. This is always just `do`
2. This is all about json patches. The actions actually modify manifests.

One could argue that the actions should be rewritten as json patches, but that didn't feel like something that needed to be done in this PR. Perhaps we should do that when we move the agent installer to the traffic-manager.

File: cmd/traffic/cmd/manager/systema.go
Comment (thallgren) [reply cmd/traffic/cmd/manager/systema.go]: So why does the linter _sometimes_ complain about this? It would be great if it captured all.

File: pkg/client/envconfig.go
Comment (thallgren) [reply pkg/client/envconfig.go]: I removed this change as it's not relevant to this PR anyway.

File: pkg/install/resource/tm_cluster_role_binding.go
Comment (thallgren) [inline pkg/install/resource/tm_cluster_role_binding.go]: This should not be hardcoded to "ambassador". A fix will arrive in subsequent commits.

File: pkg/install/tls.go
Comment (thallgren) [inline pkg/install/tls.go]: The hardcoded "ambassador" will be replaced in a subsequent commit.

File: pkg/install/resource/instance.go
Comment (LukeShu) [inline pkg/install/resource/instance.go]: CRUD, not CRDU! :P

File: pkg/install/resource/instance.go
Comment (LukeShu) [inline pkg/install/resource/instance.go]: Those should just be `scopeKey{}`, not `&scopeKey{}`.

Honestly, I'm shocked it seems to work at all with the `&`.

File: CHANGELOG.md
Comment (LukeShu) [inline CHANGELOG.md]: So I didn't actually finish writing that sentence.

File: pkg/install/resource/tm_deployment.go
Comment (LukeShu) [inline pkg/install/resource/tm_deployment.go]: So this is a global variable with state.  Which isn't a blocker, but... globals are a smell.

File: pkg/install/resource/tm_namespace.go
Comment (LukeShu) [inline pkg/install/resource/tm_namespace.go]: Can we change these from `var` to `const`?

File: pkg/install/resource/tm_svc.go
Comment (LukeShu) [inline pkg/install/resource/tm_svc.go]: Likewise, can this be `const`?

File: pkg/install/resource/tm_deployment.go
Comment (LukeShu) [reply pkg/install/resource/tm_deployment.go]: Maybe we avoid that by having `GetTrafficManagerResources()` say `&tmDeployment{}` instead of `TrafficManagerDeployment`?

Comment (thallgren) [reply pkg/install/resource/tm_deployment.go]: True that. Bad design. I'll fix it. There's no need for this one to be a variable.

File: pkg/install/resource/tm_namespace.go
Comment (thallgren) [reply pkg/install/resource/tm_namespace.go]: No, because they are of type `Instance` which is an interface. We could of course cover that with an anonymous var instead.

File: pkg/install/resource/instance.go
Comment (thallgren) [reply pkg/install/resource/instance.go]: Not intended. I'll remove that ampersand. IIRC, it works but the behavior is undefined in the spec.

File: cmd/traffic/cmd/manager/internal/mutator/service.go
Comment (LukeShu) [inline cmd/traffic/cmd/manager/internal/mutator/service.go]: If you use `dhttp` instead of the old deprecated `dutil`, then you don't have to configure `ErrorLog`.

File: cmd/traffic/cmd/manager/internal/mutator/service.go
Comment (LukeShu) [inline cmd/traffic/cmd/manager/internal/mutator/service.go]: You can type `"kube-node-lease"` as `corev1.NamespaceNodeLease` if you want.

File: CHANGELOG.md
Comment (thallgren) [reply CHANGELOG.md]: I'll replace the `TODO` with `the extended time it takes to apply everything needed for the mutator webhook.`

File: pkg/install/resource/tm_deployment.go
Comment (LukeShu) [inline pkg/install/resource/tm_deployment.go:99]: You added this in "Move install, update, and delete of traffic-manager to resource package" but probably didn't mean to add it until "Add a mutator webhook service to the traffic-manager".  You don't need to change it, but I got thrown through a loop for a minute.

File: pkg/install/resource/tm_namespace.go
Comment (thallgren) [reply pkg/install/resource/tm_namespace.go]: I'll replace with:
```
const TrafficManagerNamespace = nsResource(0)
const TrafficManagerNamespaceKeep = nsResource(1)

var _ Instance = TrafficManagerNamespace
```
and do the same thing for the other integer base types.

File: cmd/traffic/cmd/manager/internal/mutator/agent_injector.go
Comment (LukeShu) [inline cmd/traffic/cmd/manager/internal/mutator/agent_injector.go:55]: That... can't be right?  The Pod name is `GenerateName` plus a random suffix.
```
$ kubectl get pod -o yaml verylargejavaservice-6b696cb89f-blfdd |y2j|jq '{name: .metadata.name, generateName: .metadata.generateName}'
{
  "name":         "verylargejavaservice-6b696cb89f-blfdd",
  "generateName": "verylargejavaservice-6b696cb89f-"
}
```

File: pkg/install/resource/injector_svc.go
Comment (LukeShu) [inline pkg/install/resource/injector_svc.go:45]: Is there a reason we don't just shove the port in to the existing manager service? Is there a reason to choose one approach over the other?

**Review by LukeShu — CHANGES_REQUESTED** [review]
That was _so much_ easier to review, thank you!

I have 2 minor changes to request, and a question I want clarified.

- Change requested: the context key change
- Change requested: the changelog TODO
- Question: The Pod generateName thing.  Is it OK because the podName only gets used for user-messages?

Other than that, looks solid!  I mentioned a number of nitpicks and minor suggestions that you can act on or ignore.

**Comment by LukeShu** [comment]
After you make the changes, feel free to merge without waiting for me to re-review.

File: cmd/traffic/cmd/manager/internal/mutator/agent_injector.go
Comment (thallgren) [reply cmd/traffic/cmd/manager/internal/mutator/agent_injector.go:55]: Copy pasta from old impl. I'll remove it.

File: pkg/install/resource/injector_svc.go
Comment (thallgren) [reply pkg/install/resource/injector_svc.go:45]: The webhook mechanism is hardcoded to port 443 and I didn't want to dedicate that port for this since it uses a generated TLS that is useless for other purposes. The traffic-manager service has `ClusterIP: None` so it's headless and cannot do proxying and I wanted a service that proxies 443 to  8443 in the manager. I'm sure there are other ways to do this. Let's revisit.

**Review by LukeShu — APPROVED** [review]
Hit it!

**Pull request merged**
